### PR TITLE
Add an os_telemetry_collector service

### PR DESCRIPTION
**Pull request by ityaptin**
Service os_telemetry_collector implements receiving Сeilometer samples
from RabbitMQ and pulling them to InfluxDB (samples) and
ElasticSearch (resources).

### CODE REVIEW

**Comment by cloudlab-autobuild** [comment]
Can one of the admins verify this patch?

File: heka/_service.sls [hunk 1]
```diff
@@ -112,14 +112,24 @@ heka_{{ service_name }}_service:
     'splitter': {},
     'encoder': {},
     'output': {},
+  },
+  'os_telemetry_collector': {
+    'decoder': {},
+    'input': {},
+    'trigger': {},
+    'alarm_cluster': {},
```
Comment (elemoine) [inline heka/_service.sls]: You can remove `trigger` and `alarm_cluster`, as no alarms will be configured in the telemetry collector.

File: heka/files/heka.service [hunk 2]
```diff
@@ -34,6 +34,7 @@ respawn
 pre-start script
     touch /var/log/{{ service_name }}.log
     chown heka:heka /var/log/{{ service_name }}.log
+    chown -R heka:heka /var/cache/{{ service_name }}
```
Comment (elemoine) [inline heka/files/heka.service]: Why did you add this? Have you found an actual problem? If so I think a separate PR should be created.

File: heka/files/lua/common/extra_fields.lua [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+local M = {}
+setfenv(1, M) -- Remove external access to contain everything in the module
+
+environment_id = 42
+
+tags = {
+    environment_id=environment_id
+}
+
+return M
```
Comment (elemoine) [inline heka/files/lua/common/extra_fields.lua]: Remove this file. We already have an `extra_fields.lua`. It's in `heka/files` and its content is generated based on some variable in the model.

File: heka/map.jinja [hunk 4]
```diff
@@ -76,3 +76,11 @@ RedHat:
     'poolsize': 100,
   }
 }, merge=salt['pillar.get']('heka:aggregator')) %}
+
+{% set os_telemetry_collector = salt['grains.filter_by']({
+  'default': {
+    'influxdb_port': default_influxdb_port,
+    'influxdb_timeout': default_influxdb_timeout,
+    'elasticsearch_port': default_elasticsearch_port,
```
Comment (elemoine) [inline heka/map.jinja:89]: What about rabbit?

File: heka/meta/heka.yml [hunk 5]
```diff
@@ -425,3 +427,88 @@ aggregator:
...
+      queue_auto_delete: false
+{%- endif %}
+  filter:
+{%- if os_telemetry_collector.influxdb_host is defined %}
+    ceilometer_influxdb_accumulator:
+      engine: sandbox
+      module_file: /usr/share/lma_collector/filters/influxdb_accumulator.lua
+      module_dir: /usr/share/lma_collector/common;/usr/share/heka/lua_modules
+      preserve_data: false
+      message_matcher: "Fields[aggregator] == NIL && Type =~ /ceilometer_samples$/"
```
Comment (elemoine) [inline heka/meta/heka.yml]: Remove `Fields[aggregator] == NIL`. It was removed for the other services.

**Review by elemoine — CHANGES_REQUESTED** [review]
It looks good @ityaptin!

File: heka/os_telemetry_collector.sls [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{%- if pillar.heka.os_telemetry_collector is defined %}
+
+include:
+- heka._common
+
```
Comment (SwannCroiset) [inline heka/os_telemetry_collector.sls]: you need to add this : 

```
{%- from "heka/map.jinja" import os_telemetry_collector with context %}
{%- set server = os_telemetry_collector %}
```

https://github.com/tcpcloud/salt-formula-heka/blob/stacklight/heka/metric_collector.sls#L6-L7

File: heka/meta/heka.yml [hunk 7]
```diff
@@ -425,3 +427,88 @@ aggregator:
...
+      ticker_interval: 1
+      config:
+        time_precision: "ns"
+        payload_name: 'sample_data'
+        flush_count: 500
+        bulk_metric_type_matcher: 'ceilometer_samples'
+{%- endif %}
+{%- if os_telemetry_collector.influxdb_host is defined %}
+  encoder:
```
Comment (SwannCroiset) [inline heka/meta/heka.yml:482]: you'll get a wrong syntax influxdb_host is not defined but ES is defined:
`encoder:`  key will be missing

File: heka/files/heka.service (hunk 2 above)
Comment (SwannCroiset) [reply heka/files/heka.service]: agree

**Review by SwannCroiset — CHANGES_REQUESTED** [review]

File: heka/_service.sls (hunk 1 above)
Comment (ityaptin) [reply heka/_service.sls]: Ok, will do

File: heka/files/heka.service (hunk 2 above)
Comment (ityaptin) [reply heka/files/heka.service]: Ok, will remove from this PR

File: heka/files/lua/common/extra_fields.lua (hunk 3 above)
Comment (ityaptin) [reply heka/files/lua/common/extra_fields.lua]: I will remove it

File: heka/map.jinja (hunk 4 above)
Comment (ityaptin) [reply heka/map.jinja:89]: I've added a rabbit defaults here

File: heka/os_telemetry_collector.sls (hunk 6 above)
Comment (ityaptin) [reply heka/os_telemetry_collector.sls]: fixed

**Review by ityaptin — COMMENTED** [review]
@elemoine @SwannCroiset I fixed your comments

**Comment by cznewt** [comment]
I have one bigger thought, I'd rather have remote_collector implemented on ceilometer service metadata populating the remote_collector with inputs/decoder/filter definitions instead of separate heka service for os/ceilometer telemetry.

File: heka/map.jinja [hunk 8]
```diff
@@ -76,3 +81,15 @@ RedHat:
...
+    'rabbit_port': default_rabbit_port,
+    'rabbit_vhost': default_rabbit_vhost,
+    'rabbit_vhost': default_rabbit_queue,
```
Comment (simonpasquier) [inline heka/map.jinja]: typo: 'rabbitmq_vhost'

File: heka/meta/heka.yml [hunk 9]
```diff
@@ -425,3 +427,86 @@ aggregator:
...
+{%- if os_telemetry_collector.influxdb_host is defined %}
+    ceilometer_influxdb_accumulator:
+      engine: sandbox
+      module_file: /usr/share/lma_collector/filters/influxdb_accumulator.lua
+      module_dir: /usr/share/lma_collector/common;/usr/share/heka/lua_modules
+      preserve_data: false
+      message_matcher: "Type =~ /ceilometer_samples$/"
+      ticker_interval: 1
+      config:
+        time_precision: "ns"
```
Comment (simonpasquier) [inline heka/meta/heka.yml]: ideally it shouldn't be hardcoded

**Review by simonpasquier — CHANGES_REQUESTED** [review]

File: heka/map.jinja (hunk 8 above)
Comment (ityaptin) [reply heka/map.jinja]: I'll fix it

File: heka/meta/heka.yml (hunk 9 above)
Comment (ityaptin) [reply heka/meta/heka.yml]: Simon, I think there is no problem to move time_precision to variable. I will check and change it.

**Comment by ityaptin** [comment]
@cznewt: Hi! We implemented os_telemetry_collector as a separate service because in my mind new service should have availability to have own configs and been installed on different nodes. Own configs are especially important in deployments with a detached rabbitmq server for monitoring data from ceilometer (it's in a plan how I know). Also managing of this service separately could be useful for end users of ceilometer. 
Next important point is a time. I think it's better to keep this stuff in os_telemetry_collector in this time slot, because we tested it a little during development, but we could think about moving code to remote_collector in the future.

**Comment by elemoine** [comment]
> I have one bigger thought, I'd rather have remote_collector implemented on ceilometer service metadata populating the remote_collector with inputs/decoder/filter definitions instead of separate heka service for os/ceilometer telemetry.

Why? What is the problem with having a separate service for Telemetry? Having a separate service is more flexible, for example allowing the remote_collector and telemetry_collector to run on separate servers (as already mentioned by @ityaptin). So for now, I'd go with a separate instance. We can revisit this later if it causes problem.

File: heka/init.sls [hunk 10]
```diff
@@ -12,4 +12,7 @@ include:
 {%- if pillar.heka.aggregator is defined %}
 - heka.aggregator
 {%- endif %}
+{%- if pillar.heka.os_telemetry_collector is defined %}
```
Comment (simonpasquier) [inline heka/init.sls]: please move it before `- heka.aggregator`. This is because of grains file's generation.

File: heka/meta/heka.yml [hunk 11]
```diff
@@ -425,3 +427,87 @@ aggregator:
...
 {%- endif %}
+os_telemetry_collector:
+  decoder:
+    sample:
+      engine: sandbox
+      module_file: /usr/share/lma_collector/decoders/metering.lua
+      module_dir: /usr/share/lma_collector/common;/usr/share/heka/lua_modules
+      config:
+        decoder: 'ceilometer'
+      {%- if os_telemetry_collector.resource_decoding is defined %}
```
Comment (simonpasquier) [inline heka/meta/heka.yml]: the `resource_decoding` parameter can be added to map.jinja and then:
```
  decode_resources: {{ os_telemetry_collector.resource_decoding|lower }}
```

**Review by simonpasquier — CHANGES_REQUESTED** [review]

File: heka/init.sls (hunk 10 above)
Comment (ityaptin) [reply heka/init.sls]: I will do

**Review by cznewt — APPROVED** [review]
LGTM, the separate role for ceilometer service makes sense

**Review by simonpasquier — APPROVED** [review]

**Review by SwannCroiset — APPROVED** [review]

File: heka/_service.sls [hunk 12]
```diff
@@ -112,6 +112,14 @@ heka_{{ service_name }}_service:
     'splitter': {},
     'encoder': {},
     'output': {},
+  },
+  'os_telemetry_collector': {
+    'decoder': {},
```
Comment (cznewt) [inline heka/_service.sls:117]: here probably to ceilometer_collector as well

**Review by cznewt — APPROVED** [review]

**Pull request merged**